File: packages/fly-brain/src/genome.ts
// Connectome GENOME + genetic operators — the deterministic core of murmur's breeding market.
//
// WHAT A GENOME IS. A connectome built by buildConnectome(seed, opts) is fully reproducible (integer
// mulberry32 PRNG ⇒ bit-identical on every JS engine). So the COMPLETE heritable identity of one brain
// is just its EFFECTIVE generator parameters: (seed, per-layer counts, density). We call that tuple a
// `Genome`. It is small, JSON-serialisable, and — crucially — anyone can rebuild the exact brain from
// it offline (buildFromGenome) and re-derive its quantised structural spec (specFromGenome), exactly
// like the brain-manifest replay does for the base population.
//
// BREEDING. mutateGenome / crossoverGenome are PURE functions of (parents, rngSeed): the same parents +
// the same integer rngSeed always yield the same offspring genome. That determinism is what lets an
// offspring be committed on-chain by hash and re-derived by a stranger (see ConnectomeLineage.sol and
// the worker's /lineage endpoints). No randomness is hidden: the rngSeed is part of the lineage record.
//
// This module is PURE and SYNCHRONOUS (no crypto, no I/O) so it runs identically in a Worker, in Node
// tests, in a browser and in an offline CLI — mirroring manifest.ts's discipline. Hashing a genome
// (sha256 of canonicalGenome) lives in the worker (breed.ts), which has crypto.subtle.

import { buildConnectome, type ConnectomeOptions } from "./connectome.js";
import {
  connectomeStructuralSpec,
  effectiveConnectomeOptions,
  type ConnectomeStructuralSpec,
} from "./manifest.js";
import type { Connectome } from "./types.js";

/** Bump when the genome field set / operator semantics change (invalidates comparability). */
export const GENOME_SCHEMA_VERSION = 1;

/**
 * The complete heritable identity of one connectome: the effective generator parameters. Two genomes
 * that are equal produce equal brains; different genomes produce different brains.
 */
export interface Genome {
  v: number;
  /** uint32 PRNG seed — the wiring identity. */
  seed: number;
  nSensory: number;
  nInterL1: number;
  nInterL2: number;
  nModulatory: number;
  nMotorPerChannel: number;
  /** Synapse density fraction (0,1]; rounded to 4dp by operators to stay canonical-stable. */
  density: number;
}

/** Sane breeding bounds so offspring stay buildable on the edge (never 0-size, never absurd). */
export const GENOME_BOUNDS = {
  nSensory: [8, 2000],
  nInterL1: [8, 4000],
  nInterL2: [8, 4000],
  nModulatory: [4, 2000],
  nMotorPerChannel: [1, 500],
  density: [0.0005, 0.2],
} as const;

const SIZE_FIELDS = ["nSensory", "nInterL1", "nInterL2", "nModulatory", "nMotorPerChannel"] as const;
type SizeField = (typeof SIZE_FIELDS)[number];

/** Integer-only mulberry32 (bit-identical across engines). Local copy: genome.ts must stay dependency-free. */
function mulberry32(a: number): () => number {
  let t = a >>> 0;
  return () => {
    t = (t + 0x6d2b79f5) >>> 0;
    let r = Math.imul(t ^ (t >>> 15), 1 | t);
    r = (r + Math.imul(r ^ (r >>> 7), 61 | r)) ^ r;
    return ((r ^ (r >>> 14)) >>> 0) / 4294967296;
  };
}

function clampInt(x: number, lo: number, hi: number): number {
  if (!Number.isFinite(x)) return lo;
  return Math.max(lo, Math.min(hi, Math.floor(x)));
}
function clamp(x: number, lo: number, hi: number): number {
  return x < lo ? lo : x > hi ? hi : x;
}
function round4(x: number): number {
  return Math.round(x * 1e4) / 1e4;
}

/** Recursive sorted-key JSON (canonical form). Mirrors the worker's provenance canonical, locally. */
function canonicalJson(v: unknown): string {
  if (v === null || typeof v !== "object") return JSON.stringify(v);
  if (Array.isArray(v)) return "[" + v.map(canonicalJson).join(",") + "]";
  const o = v as Record<string, unknown>;
  const keys = Object.keys(o).sort();
  return "{" + keys.map((k) => JSON.stringify(k) + ":" + canonicalJson(o[k])).join(",") + "}";
}

/** The byte-stable form a genome hash is taken over: canonicalGenome(g). Deterministic. */
export function canonicalGenome(g: Genome): string {
  return canonicalJson(g);
}

/** Lift generator opts (defaults filled) into a Genome. The base population's genomes come from here. */
export function genomeFromOptions(opts: ConnectomeOptions = {}): Genome {
  const e = effectiveConnectomeOptions(opts);
  return {
    v: GENOME_SCHEMA_VERSION,
    seed: e.seed >>> 0,
    nSensory: e.nSensory,
    nInterL1: e.nInterL1,
    nInterL2: e.nInterL2,
    nModulatory: e.nModulatory,
    nMotorPerChannel: e.nMotorPerChannel,
    density: round4(e.density),
  };
}

/** Convenience: a genome for one seed over shared sizing opts. */
export function genomeFromSeed(seed: number, opts: ConnectomeOptions = {}): Genome {
  return genomeFromOptions({ ...opts, seed });
}

/**
 * POINT MUTATION: reseed the wiring and perturb exactly one layer count (± up to ~15%), with a small
 * chance of density drift. Pure in (parent, rngSeed).
 */
export function mutateGenome(parent: Genome, rngSeed: number): Genome {
  const rng = mulberry32(rngSeed >>> 0);
  const child: Genome = { ...parent, v: GENOME_SCHEMA_VERSION };
  child.seed = (parent.seed ^ Math.floor(rng() * 4294967296)) >>> 0;
  const f: SizeField = SIZE_FIELDS[Math.floor(rng() * SIZE_FIELDS.length)];
  const step = 1 + Math.floor(rng() * Math.max(1, Math.round(parent[f] * 0.15)));
  const delta = (rng() < 0.5 ? -1 : 1) * step;
  child[f] = clampInt(parent[f] + delta, GENOME_BOUNDS[f][0], GENOME_BOUNDS[f][1]);
  if (rng() < 0.5) {
    child.density = round4(
      clamp(parent.density + (rng() < 0.5 ? -0.001 : 0.001), GENOME_BOUNDS.density[0], GENOME_BOUNDS.density[1]),
    );
  }
  return child;
}

/**
 * UNIFORM CROSSOVER: each field is inherited from either parent (coin-flip per field via rngSeed); the
 * wiring seed is inherited whole from one parent (recombination, not blending). Pure in (a, b, rngSeed).
 */
export function crossoverGenome(a: Genome, b: Genome, rngSeed: number): Genome {
  const rng = mulberry32(rngSeed >>> 0);
  const pick = <T>(x: T, y: T): T => (rng() < 0.5 ? x : y);
  return {
    v: GENOME_SCHEMA_VERSION,
    seed: pick(a.seed, b.seed),
    nSensory: pick(a.nSensory, b.nSensory),
    nInterL1: pick(a.nInterL1, b.nInterL1),
    nInterL2: pick(a.nInterL2, b.nInterL2),
    nModulatory: pick(a.nModulatory, b.nModulatory),
    nMotorPerChannel: pick(a.nMotorPerChannel, b.nMotorPerChannel),
    density: round4(pick(a.density, b.density)),
  };
}

/** Rebuild the exact connectome a genome describes (deterministic). */
export function buildFromGenome(g: Genome): Connectome {
  return buildConnectome({
    seed: g.seed,
    nSensory: g.nSensory,
    nInterL1: g.nInterL1,
    nInterL2: g.nInterL2,
    nModulatory: g.nModulatory,
    nMotorPerChannel: g.nMotorPerChannel,
    density: g.density,
  });
}

/** The quantised, ULP-safe structural spec of the brain a genome describes (the replayable identity). */
export function specFromGenome(g: Genome): ConnectomeStructuralSpec {
  return connectomeStructuralSpec(buildFromGenome(g));
}

File: packages/fly-brain-rs/src/genome.rs
//! Port of genome.ts — `Genome`, `mutateGenome`, `crossoverGenome`, `canonicalGenome`,
//! `genomeFromOptions`, `genomeToConnectomeOptions`, `buildFromGenome`, `estimateConnectomeSize`.

use crate::connectome::{build_connectome, Connectome, ConnectomeOptions};
use crate::jsmath;
use crate::prng::Mulberry32;
use serde::{Deserialize, Serialize};

pub const GENOME_SCHEMA_VERSION: i64 = 1;

/// Field names follow the TS JSON shape exactly so genome files round-trip verbatim.
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Genome {
    pub v: i64,
    pub seed: u32,
    #[serde(rename = "nSensory")]
    pub n_sensory: i64,
    #[serde(rename = "nInterL1")]
    pub n_inter_l1: i64,
    #[serde(rename = "nInterL2")]
    pub n_inter_l2: i64,
    #[serde(rename = "nModulatory")]
    pub n_modulatory: i64,
    #[serde(rename = "nMotorPerChannel")]
    pub n_motor_per_channel: i64,
    pub density: f64,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum SizeField {
    NSensory,
    NInterL1,
    NInterL2,
    NModulatory,
    NMotorPerChannel,
}

const SIZE_FIELDS: [SizeField; 5] = [
    SizeField::NSensory,
    SizeField::NInterL1,
    SizeField::NInterL2,
    SizeField::NModulatory,
    SizeField::NMotorPerChannel,
];

fn bounds(f: SizeField) -> (f64, f64) {
    match f {
        SizeField::NSensory => (8.0, 2000.0),
        SizeField::NInterL1 => (8.0, 4000.0),
        SizeField::NInterL2 => (8.0, 4000.0),
        SizeField::NModulatory => (4.0, 2000.0),
        SizeField::NMotorPerChannel => (1.0, 500.0),
    }
}
pub const DENSITY_BOUNDS: (f64, f64) = (0.0005, 0.2);

impl Genome {
    fn get(&self, f: SizeField) -> i64 {
        match f {
            SizeField::NSensory => self.n_sensory,
            SizeField::NInterL1 => self.n_inter_l1,
            SizeField::NInterL2 => self.n_inter_l2,
            SizeField::NModulatory => self.n_modulatory,
            SizeField::NMotorPerChannel => self.n_motor_per_channel,
        }
    }
    fn set(&mut self, f: SizeField, x: i64) {
        match f {
            SizeField::NSensory => self.n_sensory = x,
            SizeField::NInterL1 => self.n_inter_l1 = x,
            SizeField::NInterL2 => self.n_inter_l2 = x,
            SizeField::NModulatory => self.n_modulatory = x,
            SizeField::NMotorPerChannel => self.n_motor_per_channel = x,
        }
    }
}

fn clamp_int(x: f64, lo: f64, hi: f64) -> f64 {
    if !x.is_finite() {
        return lo;
    }
    jsmath::max(lo, jsmath::min(hi, x.floor()))
}
#[inline]
fn clamp(x: f64, lo: f64, hi: f64) -> f64 {
    if x < lo {
        lo
    } else if x > hi {
        hi
    } else {
        x
    }
}
#[inline]
pub fn round4(x: f64) -> f64 {
    jsmath::round(x * 1e4) / 1e4
}

/// `canonicalGenome(g)`: sorted-key JSON with JS number formatting.
pub fn canonical_genome(g: &Genome) -> String {
    format!(
        "{{\"density\":{},\"nInterL1\":{},\"nInterL2\":{},\"nModulatory\":{},\"nMotorPerChannel\":{},\"nSensory\":{},\"seed\":{},\"v\":{}}}",
        jsmath::number_to_string(g.density),
        g.n_inter_l1,
        g.n_inter_l2,
        g.n_modulatory,
        g.n_motor_per_channel,
        g.n_sensory,
        g.seed,
        g.v
    )
}

/// `genomeFromOptions(opts)` with every option already effective.
pub fn genome_from_options(o: &ConnectomeOptions) -> Genome {
    Genome {
        v: GENOME_SCHEMA_VERSION,
        seed: o.seed,
        n_sensory: o.n_sensory as i64,
        n_inter_l1: o.n_inter_l1 as i64,
        n_inter_l2: o.n_inter_l2 as i64,
        n_modulatory: o.n_modulatory as i64,
        n_motor_per_channel: o.n_motor_per_channel as i64,
        density: round4(o.density),
    }
}

pub fn genome_to_connectome_options(g: &Genome) -> ConnectomeOptions {
    ConnectomeOptions {
        seed: g.seed,
        n_sensory: g.n_sensory.max(0) as usize,
        n_inter_l1: g.n_inter_l1.max(0) as usize,
        n_inter_l2: g.n_inter_l2.max(0) as usize,
        n_modulatory: g.n_modulatory.max(0) as usize,
        n_motor_per_channel: g.n_motor_per_channel.max(0) as usize,
        density: g.density,
    }
}

pub fn build_from_genome(g: &Genome) -> Connectome {
    build_connectome(&genome_to_connectome_options(g))
}

/// POINT MUTATION — pure in (parent, rngSeed). Draw order: seed xor, field pick, step, sign,
/// density coin, density direction.
pub fn mutate_genome(parent: &Genome, rng_seed: u32) -> Genome {
    let mut rng = Mulberry32::new(rng_seed);
    let mut child = parent.clone();
    child.v = GENOME_SCHEMA_VERSION;
    // Math.floor(rng() * 4294967296) is exactly the raw u32 draw.
    child.seed = parent.seed ^ rng.next_u32();
    let f = SIZE_FIELDS[(rng.next() * SIZE_FIELDS.len() as f64).floor() as usize];
    let pf = parent.get(f) as f64;
    let step = 1.0 + (rng.next() * jsmath::max(1.0, jsmath::round(pf * 0.15))).floor();
    let delta = (if rng.next() < 0.5 { -1.0 } else { 1.0 }) * step;
    let (lo, hi) = bounds(f);
    child.set(f, clamp_int(pf + delta, lo, hi) as i64);
    if rng.next() < 0.5 {
        let d = parent.density + (if rng.next() < 0.5 { -0.001 } else { 0.001 });
        child.density = round4(clamp(d, DENSITY_BOUNDS.0, DENSITY_BOUNDS.1));
    }
    child
}

/// UNIFORM CROSSOVER — pure in (a, b, rngSeed). One coin per field, in field order.
pub fn crossover_genome(a: &Genome, b: &Genome, rng_seed: u32) -> Genome {
    let mut rng = Mulberry32::new(rng_seed);
    let seed = if rng.next() < 0.5 { a.seed } else { b.seed };
    let n_sensory = if rng.next() < 0.5 { a.n_sensory } else { b.n_sensory };
    let n_inter_l1 = if rng.next() < 0.5 { a.n_inter_l1 } else { b.n_inter_l1 };
    let n_inter_l2 = if rng.next() < 0.5 { a.n_inter_l2 } else { b.n_inter_l2 };
    let n_modulatory = if rng.next() < 0.5 { a.n_modulatory } else { b.n_modulatory };
    let n_motor_per_channel = if rng.next() < 0.5 { a.n_motor_per_channel } else { b.n_motor_per_channel };
    let density = round4(if rng.next() < 0.5 { a.density } else { b.density });
    Genome {
        v: GENOME_SCHEMA_VERSION,
        seed,
        n_sensory,
        n_inter_l1,
        n_inter_l2,
        n_modulatory,
        n_motor_per_channel,
        density,
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct SizeEstimate {
    pub neurons: i64,
    pub synapses: i64,
}

fn channel_count(n: i64, r: i64, m: i64) -> i64 {
    if n <= r {
        return 0;
    }
    ((n - r - 1) as f64 / m as f64).floor() as i64 + 1
}

/// `estimateConnectomeSize(g)` — closed-form, no build.
pub fn estimate_connectome_size(g: &Genome) -> SizeEstimate {
    let n_sens = g.n_sensory.max(0);
    let n_l1 = g.n_inter_l1.max(0);
    let n_l2 = g.n_inter_l2.max(0);
    let n_mod = g.n_modulatory.max(0);
    let n_motor_per = g.n_motor_per_channel.max(0);
    let d = jsmath::max(0.0, g.density);

    let l2_half = (n_l2 as f64 / 2.0).floor() as i64;
    let l2_right = n_l2 - l2_half;
    let neurons = n_sens + n_l1 + n_l2 + n_mod + 5 * n_motor_per;

    let fan = |from_size: i64, dens: f64| -> i64 { ((from_size as f64 * dens).floor() as i64).max(1) };

    let mut syn: i64 = 0;
    syn += n_l1 * fan(n_sens, d * 1.5);
    syn += n_l2 * fan(n_l1, d * 1.2);
    syn += l2_right * fan(l2_half, d * 0.8);
    syn += l2_half * fan(l2_right, d * 0.8);
    syn += l2_half * fan(l2_half, d * 0.3);
    syn += l2_right * fan(l2_right, d * 0.3);
    syn += n_motor_per * 40;
    syn += n_motor_per * 40;
    syn += 2 * n_motor_per * 30;
    let gus = channel_count(n_sens, 4, 10);
    syn += n_motor_per * gus.clamp(1, 6);
    syn += (0.05 * neurons as f64).ceil() as i64;
    syn += n_mod * fan(n_sens + n_l1, d * 0.4);
    let threat = channel_count(n_sens, 7, 10);
    syn += ((threat * n_mod) as f64 * 0.25).ceil() as i64;

    SizeEstimate { neurons, synapses: syn }
}
